lib/kernel.js

```diff
@@ -17,6 +17,14 @@ import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
 
+interface KernelMiddleware {
+  interrupt: () => void;
+  shutdown: (restart: boolean, onRestarted: ?Function) => void;
+  execute: (code: string, callWatches: boolean, onResults: Function) => void;
+  complete: (code: string, onResults: Function) => void;
+  inspect: (code: string, cursorPos: number, onResults: Function) => void;
+}
```

Comment:
If you think that callbacks are too old-fashioned, it may be possible to convert these to async syntax. Note however that `onResults` gets called multiple times, so this will rely on async iterator features that are quite recent.

Reply:
I like the idea of moving to async and promises as much as possible. Though I don't know if it'd be worth the effort in that case.

I'd rather prefer us moving to [`@nteract/messaging`](https://github.com/nteract/nteract/tree/master/packages/messaging), [`enchannel-zmq-backend`](https://github.com/nteract/nteract/tree/master/packages/enchannel-zmq-backend) and [`rx-jupyter`](https://github.com/nteract/nteract/tree/master/packages/rx-jupyter) using [RxJS](http://reactivex.io/rxjs/) which provides excellent primitives for handling streaming data.
 
Alternatively we could rely more on MobX for these kind of things.